draft-irtf-cfrg-opaque.md

```diff
@@ -851,21 +851,157 @@ Steps:
+
+Prior to the execution of these functions, both the client and the server must be set to use the same parameters described
+in {{configurations}}.
+
+In order to achieve this interlacing, the AKE should implement the following interface:
+- `ke1, client_secret = AKEStart(credential_request, client_info)`
+- `ke2, client_info, expected_client_mac, session_key = AKEResponse(server_identity, server_private_key, client_identity, client_public_key, server_info, ke1, credential_response)`
+- `ke3, server_info, session_key = AKEClientFinalize(client_secret, client_identity, client_private_key, server_identity, server_public_key, ke1, ke2)`
+- `valid = AKEServerFinalize(expected_client_mac, KE3)`
+
+The `credential_request`, `client_public_key`, `credential_response`, `client_private_key`, and `server_public_key`
+(in `AKEClientFinalize()`) values are produced by the credential retrieval process as described in {{credential-retrieval}}.
+
+These functions and values are particularly suited for OPAQUE-3DH, and might need adaptation if another AKE is used.
+
+### OPAQUE Client {#opaque-client}
+
+~~~
+OPAQUEStart(client_identity, password, client_info)
+
+Input:
+- client_identity, the optional encoded client identity, which is nil if not specified
+- password, an opaque byte string containing the client's password
+- client_info, optional application-specific information to exchange during the protocol
+
+Output:
+- KE1, a KE1 message structure
+- blind, the OPRF blinding scalar
+- client_secret, the client's Diffie-Hellman secret share for the session
+
+Steps:
+1. request, blind = CreateCredentialRequest(password)
+2. ke1, client_secret = AKEStart(request, client_info)
+3. Output (ke1, blind, client_secret)
+~~~
+
+~~~
+OPAQUEClientFinalize(password, blind, client_secret, client_identity, server_identity, ke1, ke2)
+
+Input:
+- password, an opaque byte string containing the client's password
+- blind, the OPRF blinding scalar
+- client_secret, the client's Diffie-Hellman secret share for the session
+- client_identity, the optional encoded client identity, which is set to its public key if not specified
+- server_identity, the optional encoded server identity, which is set to its public key if not specified
+- ke1, a KE1 message structure
+- ke2, a KE2 message structure
+
+Output:
+- ke3, a KE3 message structure
+- server_info, optional application-specific information sent by the server
+- session_key, the session's shared secret 
+
+Steps:
+1. Create Credentials creds with (client_identity, server_identity)
+2. client_private_key, server_public_key, export_key = RecoverCredentials(password, blind, ke2.CredentialResponse)
+3. ke3, server_info, session_key = AKEClientFinalize(client_secret, client_identity, client_private_key,
+                                               server_identity, server_public_key, ke1, ke2)
+4. Output (ke3, server_info, session_key)
+~~~
+
+The client's state between the two functions, except the parameters, are:
```

Comment:
As above, I would recommend folding all this into a "state" or "context" object.